### PR TITLE
Init structure + init data

### PR DESCRIPTION
**Pull request by joYyHack**

### CODE REVIEW

File: .vscode/settings.json
Comment (Arvolear) [inline .vscode/settings.json:1]: Please gitignore this

File: circuits/Smt.circom
Comment (Arvolear) [inline circuits/Smt.circom:1]: Leave only the SMT circuit

File: contracts/SmtChecker.sol
Comment (Arvolear) [inline contracts/SmtChecker.sol:1]: Do you actually need this for testing?

File: contracts/libs/SMT.sol
Comment (Arvolear) [inline contracts/libs/SMT.sol:1]: Use the SMT from Solarity

File: zk-out/tau/powersOfTau28_hez_final_08.ptau
Comment (Arvolear) [inline zk-out/tau/powersOfTau28_hez_final_08.ptau:1]: gitignore these

File: zk-out/Sum2/circuit_0000.zkey
Comment (Arvolear) [inline zk-out/Sum2/circuit_0000.zkey:1]: Are these used?

File: test/Smt.test.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  });
+
+  afterEach(async () => {
+    await reverter.revert();
+  });
+
+  it("should prove the tree", async () => {
+    let leaves: string[] = [];
+
+    for (let i = 0; i < 10; i++) {
```
Comment (Arvolear) [inline test/Smt.test.ts:56]: Cool, but we need more tests

File: contracts/SmtChecker.sol
Comment (joYyHack) [reply contracts/SmtChecker.sol:1]: check again please, a bit modified file and renamed to "SmtMock"

File: test/Smt.test.ts (hunk 1 above)
Comment (joYyHack) [reply test/Smt.test.ts:50]: Will be done by Ruslan as he already works on SMT tree. I will take IMT

File: zk-out/Sum2/circuit_0000.zkey
Comment (joYyHack) [reply zk-out/Sum2/circuit_0000.zkey:1]: actually no. Ignored all .zkey, except final.zkey

File: contracts/SmtChecker.sol
Comment (joYyHack) [reply contracts/SmtChecker.sol:1]: Check it again, please. Slightly modified the file and renamed it to "SmtMock"

File: zk-out/Smt/challenge_phase2_0003
Comment (Arvolear) [inline zk-out/Smt/challenge_phase2_0003:1]: We probably don't need these files

File: test/Smt.test.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      value: merkleProof.value,
+    })) as ProofStruct;
+
+    let calldata = await snarkjs.groth16.exportSolidityCallData(
+      proofStruct.proof,
+      proofStruct.publicSignals
+    );
+
+    calldata = JSON.parse(`[${calldata}]`) as Calldata;
+    const [pA, pB, pC, publicSignals] = calldata;
```
Comment (Arvolear) [inline test/Smt.test.ts:73]: Can this be moved to a helper function?

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "target": "es2020",
```
Comment (Arvolear) [inline tsconfig.json:3]:
```suggestion
    "target": "es2022",
```

File: scripts/circom.ts [hunk 4]
```diff
@@ -80,4 +91,32 @@ export async function clean() {
...
+  );
+}
+
+async function* walkOverDir(
+  dir: string
+): AsyncGenerator<string, void, unknown> {
+  for await (const d of await fs.promises.opendir(dir)) {
+    const entry = path.join(dir, d.name);
+    if (d.isDirectory()) yield* walkOverDir(entry);
+    else if (d.isFile()) yield entry;
```
Comment (Arvolear) [inline scripts/circom.ts:133]: Style

File: scripts/circom.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    await fs.promises.rm(`${outputPath}/`, { recursive: true });
+
+    logger.info("Output is cleaned");
+  }
+}
+
+async function updateCircuitConfig() {
+  config.build.circuits = [];
+  let counter = 0;
+  for await (const p of walkOverDir(circuitsDir)) {
```
Comment (Arvolear) [inline scripts/circom.ts:110]:
```suggestion
  let counter = 0;
  
  for await (const p of walkOverDir(circuitsDir)) {
```

File: tasks/tasks.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+task("circom:verifier", "Generate verifier contract")
+  .addOptionalPositionalParam<string>(
+    "circuitIds",
+    "Ids of the circuit. If not specified, verifiers will be created for each curcuit."
+  )
+  .setAction(async (args) => {
+    await createVerifier(args.circuitIds);
+  });
+
+task("circom:build", "copmpile and create verifier contract for circuits")
```
Comment (Arvolear) [inline tasks/tasks.ts:23]:
```suggestion
task("circom:build", "compile and create verifier contracts for circuits")
```

File: tasks/tasks.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  .addOptionalPositionalParam<string>(
+    "circuitIds",
+    "Ids of the circuit. If not specified, all circuites will be compiled."
+  )
+  .setAction(async (args) => {
+    console.log(args);
+    await compile(args.circuitIds);
+  });
+
+task("circom:verifier", "Generate verifier contract")
```
Comment (Arvolear) [inline tasks/tasks.ts:14]:
```suggestion
task("circom:verifier", "Generate verifier contracts")
```

File: tasks/tasks.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    "Ids of the circuit. If not specified, verifiers will be created for each curcuit."
+  )
+  .setAction(async (args) => {
+    await createVerifier(args.circuitIds);
+  });
+
+task("circom:build", "copmpile and create verifier contract for circuits")
+  .addOptionalPositionalParam<string>(
+    "circuitIds",
+    "Ids of the circuit. If not specified, verifiers will be created for each curcuit."
```
Comment (Arvolear) [inline tasks/tasks.ts:26]:
```suggestion
    "Id of a circuit. If not specified, verifiers will be created for each circuit in the project."
```

File: tasks/tasks.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  )
+  .setAction(async (args) => {
+    await build(args.circuitIds);
+  });
+
+task(
+  "circom:clean",
+  "delete all output files from zk-out folder and all verifiers"
```
Comment (Arvolear) [inline tasks/tasks.ts:34]:
```suggestion
  "Delete all output files from zk-out folder together with verifiers"
```

File: test/Smt.test.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    smtVerifier = await ethers.deployContract("SmtVerifier");
+
+    const SMTChecker = new SmtMock__factory({
+      ["@iden3/contracts/lib/Poseidon.sol:PoseidonFacade"]:
+        await poseidonFacade.getAddress(),
+    });
+
+    smtChecker = await SMTChecker.connect(owner).deploy();
+
+    reverter.snapshot();
```
Comment (Arvolear) [inline test/Smt.test.ts:38]:
```suggestion
    await reverter.snapshot();
```

File: test/Smt.test.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    const merkleProof = await smtChecker.getProof(leaves[5]);
+
+    let proofStruct = (await smtCircuit.genProof({
+      root: merkleProof.root,
+      siblings: merkleProof.siblings,
+      key: merkleProof.key,
+      value: merkleProof.value,
+    })) as ProofStruct;
+
+    const [pA, pB, pC, publicSignals] = await generateCallData(proofStruct);
```
Comment (Arvolear) [inline test/Smt.test.ts:65]:
```suggestion
    const [pA, pB, pC, publicSignals] = await generateCalldata(proofStruct);
```

File: tasks/tasks.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import { task } from "hardhat/config";
+import { compile, createVerifier, build, clean } from "../scripts/circom";
+
+task("circom:compile", "Compile circuits")
+  .addOptionalPositionalParam<string>(
+    "circuitIds",
+    "Ids of the circuit. If not specified, all circuites will be compiled."
```
Comment (Arvolear) [inline tasks/tasks.ts:7]:
```suggestion
    "Id of a circuit. If not specified, all circuits in the project will be compiled"
```

File: test/Smt.test.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    [owner] = await ethers.getSigners();
+
+    const poseidonFacade = await deployPoseidonFacade();
+
+    smtVerifier = await ethers.deployContract("SmtVerifier");
+
+    const SMTChecker = new SmtMock__factory({
+      ["@iden3/contracts/lib/Poseidon.sol:PoseidonFacade"]:
+        await poseidonFacade.getAddress(),
+    });
```
Comment (Arvolear) [inline test/Smt.test.ts:34]: Is this library linking? This is not how it is done

**Review by Arvolear — APPROVED** [review]
Let's go!

**Pull request merged**